e2e/node_delete_test.go

```diff
@@ -0,0 +1,137 @@
+package e2e
+
+import (
+	_ "embed"
+	"errors"
+	"fmt"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/topolvm/topolvm"
+	topolvmv1 "github.com/topolvm/topolvm/api/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const nsNodeDeleteTest = "node-delete-test"
+
+//go:embed testdata/node_delete/statefulset-template.yaml
+var statefulSetTemplateYAML string
+
+func testNodeDelete() {
+
+	BeforeEach(func() {
+		// Skip because this test require multiple nodes but there is just one node in daemonset lvmd test environment.
+		skipIfDaemonsetLvmd()
+
+		createNamespace(nsNodeDeleteTest)
+	})
+
+	AfterEach(func() {
+		skipIfDaemonsetLvmd()
+
+		kubectl("delete", "namespaces/"+nsNodeDeleteTest)
+	})
+
+	It("should re-create pods and PVCs on other nodes than the deleted node", func() {
+		By("waiting for node finalizer set")
+		Eventually(func() error {
+			var nodes corev1.NodeList
+			err := getObjects(&nodes, "nodes", "-l=node-role.kubernetes.io/control-plane!=")
+			if err != nil {
+				return err
+			}
+			for _, node := range nodes.Items {
+				if !controllerutil.ContainsFinalizer(&node, topolvm.GetNodeFinalizer()) {
+					return errors.New("topolvm finalizer is not attached")
+				}
+			}
+			return nil
+		}).Should(Succeed())
+
+		statefulsetName := "test-sts"
+		By("applying statefulset")
+		statefulsetYAML := []byte(fmt.Sprintf(statefulSetTemplateYAML, statefulsetName, statefulsetName))
+		_, err := kubectlWithInput(statefulsetYAML, "-n", nsNodeDeleteTest, "apply", "-f", "-")
+		Expect(err).ShouldNot(HaveOccurred())
+
+		Eventually(func() error {
+			var st appsv1.StatefulSet
+			err := getObjects(&st, "statefulset", "-n", nsNodeDeleteTest, statefulsetName)
+			if err != nil {
+				return err
+			}
+			if st.Status.ReadyReplicas != 3 {
+				return fmt.Errorf("statefulset replica is not 3: %d", st.Status.ReadyReplicas)
+			}
+			return nil
+		}).Should(Succeed())
+
+		By("getting a target pod")
+		var targetPod *corev1.Pod
+		targetNode := "topolvm-e2e-worker3"
+		var pods corev1.PodList
+		err = getObjects(&pods, "pods", "-n", nsNodeDeleteTest)
+		Expect(err).ShouldNot(HaveOccurred())
+
+		for _, pod := range pods.Items {
+			if pod.Spec.NodeName == targetNode {
+				targetPod = &pod
+				break
+			}
+		}
+		Expect(targetPod).ShouldNot(BeNil())
+
+		By("getting LVs to be cleaned up")
+		var logicalVolumeList topolvmv1.LogicalVolumeList
+		err = getObjects(&logicalVolumeList, "logicalvolumes")
+		Expect(err).ShouldNot(HaveOccurred())
+
+		var targetLVs []topolvmv1.LogicalVolume
+		for _, lv := range logicalVolumeList.Items {
+			if lv.Spec.NodeName == targetNode {
+				targetLVs = append(targetLVs, lv)
+			}
+		}
+
```

Comment:
Is it guaranteed that `targetLVs` is not empty?
Then, it would be good to check it after the loop.

Reply:
LogicalVolume is used internally, we don't have any user-visible specs. I think e2e should not test implementation detail but test user-visible specs. So, it is unnecessary to check it, IMO.